### PR TITLE
NL: Improve screenshot scraper

### PR DESCRIPTION
**Pull request by pradh**
1. Use changes from https://github.com/datacommonsorg/website/pull/2678 and https://github.com/datacommonsorg/website/pull/2679
2. Use the Selenium API instead of Beautifulsoup
3. Make recursive calls to local website (rather than dev.dc.org)
4. Avoid base64 encoding SVG (online consensus is that it is not necessary since SVG is text, and obfuscates)
5. Add integration tests (skip SVGs, they'll be in webdriver)
6. Change NL cloud build to build client libs
7. Use a driver instance per request (for reasons described in html.py)

### CODE REVIEW

File: server/routes/nl/html.py [hunk 1]
```diff
@@ -52,14 +53,17 @@ def screenshot():
 
 @bp.route('/data')
 def data_page():
+  logging.info('NL Data Page: Enter')
```
Comment (chejennifer) [inline server/routes/nl/html.py:56]: did you mean to keep these Enter and Exit logs? or were they for testing?

also in api.py

File: server/routes/nl/api.py [hunk 2]
```diff
@@ -486,8 +487,11 @@ def data():
   if not query:
     logging.info("Query was empty")
     query_detection = _detection("", "", query_detection_debug_logs, counters)
-    return _result_with_debug_info(res, "Aborted: Query was Empty.",
-                                   query_detection, escaped_context_history, {})
+    data_dcit = _result_with_debug_info(res, "Aborted: Query was Empty.",
```
Comment (chejennifer) [inline server/routes/nl/api.py:490]: data_dict?

File: server/lib/nl/scraper.py [hunk 3]
```diff
@@ -238,88 +256,88 @@ def _check_svgs_drawn(driver):
...
+        chart['svg'] = _get_inline_svg(chart_container)
+        _get_legend_data(chart_container, 'div.legend', 'a', chart)
       elif 'map-chart' in classes:
         chart['type'] = 'MAP'
-        map_area = chart_container.find('div', {'class': 'map'})
-        chart['svg'] = _to_svg(map_area.find('svg'))
-        legend_area = chart_container.find('div', {'class': 'legend'})
-        chart['legend_svg'] = _to_svg(legend_area.find('svg'))
+        map_area = chart_container.find_element(By.CSS_SELECTOR, 'div.map')
+        chart['svg'] = _get_inline_svg(map_area)
```
Comment (chejennifer) [inline server/lib/nl/scraper.py:323]: map chart should be using data-svg now right?

File: server/integration_tests/nl_screenshot_test.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  def run_test(self, test_dir, query):
+    cfg = libconfig.get_config()
+    cfg.WEBSITE_ROOT = self.get_server_url()
+    logging.info(query)
+    logging.info(self.get_server_url())
+    resp = requests.get(
+        f'{self.get_server_url()}/nl/screenshot?q={query}').json()
+    for chart in resp.get('charts', []):
+      chart['svg'] = ''
```
Comment (chejennifer) [inline server/integration_tests/nl_screenshot_test.py:77]: is this test not checking svgs? If not checking svg content, maybe should still check that resp gives non empty chart['svg']?

File: server/integration_tests/nl_screenshot_test.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      if 'data_csv' in chart:
+        lines = chart['data_csv'].splitlines()
+        sorted_lines = [lines[0]] + sorted(lines[1:])
```
Comment (chejennifer) [inline server/integration_tests/nl_screenshot_test.py:80]: with the change to make csv lines deterministic, do we still need this part?

**Review by chejennifer — COMMENTED** [review]
nice & thanks for adding tests!

File: server/integration_tests/nl_screenshot_test.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+from flask_testing import LiveServerTestCase
+import requests
+
+from nl_server.__init__ import create_app as create_nl_app
+from server.__init__ import create_app as create_web_app
+import server.lib.config as libconfig
+import server.lib.util as libutil
+
+# Explicitly set multiprocessing start method to 'fork' so tests work with
```
Comment (shifucun) [inline server/integration_tests/nl_screenshot_test.py:28]: can remove this comment block?

File: server/integration_tests/nl_screenshot_test.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+# python3.8+ on MacOS.
+# https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
+# This code must only be run once per execution.
+# if sys.version_info >= (3, 8) and sys.platform == "darwin":
+#   multiprocessing.set_start_method("fork")
+#   os.environ['no_proxy'] = '*'
+
+_dir = os.path.dirname(os.path.abspath(__file__))
+
+_NL_SERVER_URL = 'http://127.0.0.1:6060'
```
Comment (shifucun) [inline server/integration_tests/nl_screenshot_test.py:38]: can use app.config['NL_ROOT']

File: server/integration_tests/nl_screenshot_test.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+  def create_app(self):
+    """Returns the Flask Server running Data Commons."""
+    return create_web_app()
+
+  def run_test(self, test_dir, query):
+    cfg = libconfig.get_config()
+    cfg.WEBSITE_ROOT = self.get_server_url()
+    logging.info(query)
+    logging.info(self.get_server_url())
```
Comment (shifucun) [inline server/integration_tests/nl_screenshot_test.py:73]: remove logging here?

File: server/lib/nl/scraper.py [hunk 9]
```diff
@@ -80,7 +84,7 @@
...
...28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3Ccircle%20class%3D%22dot%22%20cx%3D%22375.8805855524371%22%20cy%3D%2231.487142857142857%22%20r%3D%223%22%20style%3D%22fill%3A%20rgb%2894%2C%2079%2C%20162%29%3B%20stroke%3A%20rgb%28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3Ccircle%20class%3D%22dot%22%20cx%3D%22294.39788985487405%22%20cy%3D%2231.905142857142863%22%20r%3D%223%22%20style%3D%22fill%3A%20rgb%2894%2C%2079%2C%20162%29%3B%20stroke%3A%20rgb%28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3Ccircle%20class%3D%22dot%22%20cx%3D%22212.69195389512598%22%20cy%3D%2231.716571428571427%22%20r%3D%223%22%20style%3D%22fill%3A%20rgb%2894%2C%2079%2C%20162%29%3B%20stroke%3A%20rgb%28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3Ccircle%20class%3D%22dot%22%20cx%3D%22131.20925819756297%22%20cy%3D%2230.978%22%20r%3D%223%22%20style%3D%22fill%3A%20rgb%2894%2C%2079%2C%20162%29%3B%20stroke%3A%20rgb%28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3Ccircle%20class%3D%22dot%22%20cx%3D%2249.7265625%22%20cy%3D%2231.801428571428573%22%20r%3D%223%22%20style%3D%22fill%3A%20rgb%2894%2C%2079%2C%20162%29%3B%20stroke%3A%20rgb%28255%2C%20255%2C%20255%29%3B%22%3E%3C/circle%3E%3C/g%3E%3C/g%3E%3C/svg%3E",
```
Comment (shifucun) [inline server/lib/nl/scraper.py:87]: what are these large blobs used for?

File: server/integration_tests/nl_screenshot_test.py (hunk 6 above)
Comment (pradh) [reply server/integration_tests/nl_screenshot_test.py]: Done.

File: server/integration_tests/nl_screenshot_test.py (hunk 7 above)
Comment (pradh) [reply server/integration_tests/nl_screenshot_test.py:30]: The code below explicitly sets 6060, so it perhaps makes sense to hardcode this too?

(this is copied from nl integration test)

File: server/integration_tests/nl_screenshot_test.py (hunk 8 above)
Comment (pradh) [reply server/integration_tests/nl_screenshot_test.py]: Done.

File: server/lib/nl/scraper.py (hunk 9 above)
Comment (pradh) [reply server/lib/nl/scraper.py:88]: This is the demo query used by bard that's "cached" without any backend processing.

Added a TODO to remove

File: server/lib/nl/scraper.py (hunk 3 above)
Comment (pradh) [reply server/lib/nl/scraper.py]: Just sync'd your PR.  Fixed.

File: server/integration_tests/nl_screenshot_test.py (hunk 4 above)
Comment (pradh) [reply server/integration_tests/nl_screenshot_test.py:68]: Good idea!

File: server/integration_tests/nl_screenshot_test.py (hunk 5 above)
Comment (pradh) [reply server/integration_tests/nl_screenshot_test.py]: Just sync'd and removed this.

**Review by pradh — COMMENTED** [review]
Thanks for the reviews!

File: server/routes/nl/html.py [hunk 10]
```diff
@@ -43,11 +44,33 @@ def page():
...
 
+def _get_selenium_driver():
+  options = webdriver.chrome.options.Options()
+  options.add_argument("--headless=new")
+  options.add_argument("--disable-gpu")
+  options.add_argument("--no-sandbox")
+  options.add_argument("enable-automation")
+  options.add_argument("--disable-infobars")
+  options.add_argument("--disable-dev-shm-usage")
+  return webdriver.Chrome(options=options)
```
Comment (shifucun) [inline server/routes/nl/html.py:55]: how expensive is this? Is there a delay here?

**Comment by shifucun** [comment]
Thanks for the updates! We have more understanding of complex testing now!

**Review by shifucun — APPROVED** [review]

**Review by chejennifer — APPROVED** [review]
thanks for the updates!

File: server/routes/nl/html.py (hunk 10 above)
Comment (pradh) [reply server/routes/nl/html.py:55]: Nearly a second.  Might need to explore other options soon.

**Pull request merged**